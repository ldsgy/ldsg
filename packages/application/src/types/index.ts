export * from "./handler-resource-settings";
