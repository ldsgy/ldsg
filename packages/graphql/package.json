{
  "name": "@ldsg/graphql",
  "version": "0.1.0-alpha.21",
  "description": "",
  "main": "dist/index.js",
  "files": [
    "dist/"
  ],
  "types": "dist",
  "scripts": {
    "build": "pnpm clean && tsc --project tsconfig.build.json",
    "clean": "rm -rf dist"
  },
  "keywords": [],
  "author": "wanxger",
  "license": "MIT",
  "publishConfig": {
    "access": "public"
  },
  "packageManager": "pnpm@10.7.1",
  "dependencies": {
    "@ldsg/handler": "workspace:*",
    "@ldsg/resource": "workspace:*",
    "@ldsg/resource-definition": "workspace:*",
    "graphql-compose": "^9.1.0",
    "graphql-yoga": "^5.13.4"
  },
  "devDependencies": {
    "@ldsg/application": "workspace:*",
    "@ldsg/types": "workspace:*",
    "@types/supertest": "^6.0.0",
    "graphql": "16.11.0",
    "supertest": "^6.3.4"
  }
}
