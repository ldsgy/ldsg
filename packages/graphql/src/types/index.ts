export * from "./modify-graphql-schema";
export * from "./specific-resource-settings";
