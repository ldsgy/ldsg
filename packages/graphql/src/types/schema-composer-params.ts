import { SchemaComposer } from "graphql-compose";

export interface SchemaComposerParams {
  schemaComposer: SchemaComposer;
}
