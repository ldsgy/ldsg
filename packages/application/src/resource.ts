import { Resource } from "@ldsg/resource";
import express, { Express } from "express";
import { ApplicationSpecificResourceSettings } from "./types";

type CreateApp = () => Promise<Express>;

interface AddRoutesParams {
  app: Express;
}

type AddRoutes = (params: AddRoutesParams) => Promise<void>;

export class ApplicationResource extends Resource<ApplicationSpecificResourceSettings> {
  addRoutes: AddRoutes = async (params) => {
    const { app } = params;

    const {
      settings: { name },
      getFilteredResources,
    } = this;

    const getFilteredResourcesRes = getFilteredResources({});

    const { resources } = getFilteredResourcesRes;

    resources.forEach((resource) => {
      if ("addRoute" in resource && typeof resource.addRoute === "function") {
        resource?.addRoute();
      }
    });

    app.get("/", (req, res) => {
      res.send(`Hello, ${name ?? "World"}!`);
    });
  };

  createApp: CreateApp = async () => {
    const app = express();

    await this.addRoutes({
      app,
    });

    return app;
  };
}
