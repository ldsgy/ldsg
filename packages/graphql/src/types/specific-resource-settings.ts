import { SpecificResourceSettings } from "@ldsg/types";

export interface GraphqlSpecificResourceSettings
  extends SpecificResourceSettings {
  /**
   * GraphQL Endpoint
   */
  endpoint?: string;
}
