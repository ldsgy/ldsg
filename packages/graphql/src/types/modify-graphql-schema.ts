import { SchemaComposerParams } from "./schema-composer-params";

export type ModifyGraphQLSchema = (params: SchemaComposerParams) => void;
