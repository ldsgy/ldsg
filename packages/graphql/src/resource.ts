import { ExtendExpressApp } from "@ldsg/application";
import { Resource } from "@ldsg/resource";
import { SchemaComposer } from "graphql-compose";
import { createYoga } from "graphql-yoga";
import { GraphqlSpecificResourceSettings, ModifyGraphQLSchema } from "./types";

export class GraphqlResource extends Resource<GraphqlSpecificResourceSettings> {
  extendExpressApp: ExtendExpressApp = async (params) => {
    const { app } = params;

    const {
      parentId,
      settings: { graphqlEndpoint },
      getFilteredResources,
    } = this;

    const { resources } = getFilteredResources({
      parentId,
    });

    /**
     * TODO: extends graphql schema
     */

    const yoga = createYoga({
      graphqlEndpoint,
    });

    app.use(yoga.graphqlEndpoint, yoga);
  };

  getGraphQLSchema = () => {
    const schemaComposer = new SchemaComposer();

    const { parentId, getFilteredResources } = this;

    const { resources } = getFilteredResources<{
      modifyGraphQLSchema?: ModifyGraphQLSchema;
    }>({
      parentId,
    });

    resources.forEach((resource) => {
      resource.modifyGraphQLSchema?.({
        schemaComposer,
      });
    });

    const schema = schemaComposer.buildSchema();

    const res = {
      schema,
    };

    return res;
  };
}
