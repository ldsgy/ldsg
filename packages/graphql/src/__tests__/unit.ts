import { Resource } from "@ldsg/resource";
import { printSchema } from "graphql";
import { GraphqlResource } from "../resource";
import { ModifyGraphQLSchema } from "../types";

test("unit", () => {
  const graphqlResource = new GraphqlResource({
    id: "test-graphql",
    kind: "GRAPHQL",
    parentId: "test-application",
    settings: {
      title: "",
      description: "",
      graphqlEndpoint: "/test-graphql",
    },
  });

  class AResource extends Resource {
    modifyGraphQLSchema: ModifyGraphQLSchema = (params) => {
      const { schemaComposer } = params;

      const PostTC = schemaComposer.createObjectTC({
        name: "Post",
        fields: {
          id: "Int!",
          title: "String",
          votes: "Int",
          authorId: "Int",
        },
      });

      schemaComposer.Query.addFields({
        posts: {
          type: "[Post]",
          resolve: () => [
            { id: 1, authorId: 1, title: "Introduction to GraphQL", votes: 2 },
            { id: 2, authorId: 2, title: "Welcome to Apollo", votes: 3 },
            { id: 3, authorId: 2, title: "Advanced GraphQL", votes: 1 },
            { id: 4, authorId: 3, title: "Launchpad is Cool", votes: 7 },
          ],
        },
      });
    };
  }

  new AResource({
    id: "test-a",
    kind: "A",
    parentId: "test-application",
    settings: {
      title: "",
      description: "",
    },
  });

  const { schema } = graphqlResource.getGraphQLSchema();

  expect(printSchema(schema)).toMatchSnapshot();
});
