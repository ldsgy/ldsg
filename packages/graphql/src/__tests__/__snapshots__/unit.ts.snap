// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`unit 1`] = `
"type Query {
  posts: [Post]
}

type Post {
  id: Int!
  title: String
  votes: Int
  authorId: Int
}"
`;
